File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceEvent.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/ReferenceEvent.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Event;
import com.bloomberglp.blpapi.Message;
import com.bloomberglp.blpapi.CorrelationID;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class ReferenceEvent extends Event
{
    public ReferenceEvent(ReferenceRequest request) throws Exception
    {
        super._request = request;
        super._messages = this.generateMessages();
    }
    
    private List<Message> generateMessages() throws Exception
    {
        List<Message> result = new ArrayList<Message>();
        ReferenceRequest rreq = (ReferenceRequest)super._request;

        Map<String, Map<String, Object>> securities = new HashMap<String, Map<String, Object>>();
        for(int i = 0; i < rreq.securities().size(); i++)
        {
        	String security = rreq.securities().get(i);
        	if(!securities.containsKey(security))
        	{
        		String strPattern = "[A-Z]{1,4}\\s+\\d{6}[CP]\\d{8} EQUITY";
        		Pattern p = Pattern.compile(strPattern);
        		Matcher m = p.matcher(security.toUpperCase());
        		boolean isOption = m.find();
        		
        		Map<String, Object> fieldData = new HashMap<String, Object>();
        		securities.put(security, fieldData);
        		
        		for(int j = 0; j < rreq.fields().size(); j++)
        		{
        			String field = rreq.fields().get(j);
                    Object value = com.bloomberglp.blpapi.RandomDataGenerator.referenceDataFromFieldName(field, security, isOption, rreq);
                    
                    if(value != null)
                    {
                    	fieldData.put(field, value);
                    }
        		}
        		
        	}
        }

        CorrelationID corr = super._request.correlationId(); 
        com.bloomberglp.blpapi.ReferenceMessage msg = new com.bloomberglp.blpapi.ReferenceMessage(corr, securities);
        result.add(msg);
        
        return result;
        
    }
}

File: BEmu/ReferenceDataRequest/ReferenceEvent.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="ReferenceDataRequest/ReferenceEvent.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using System.Text.RegularExpressions;

    internal class ReferenceEvent : Event
    {
        private readonly List<Message> _messages;

        internal ReferenceEvent(ReferenceRequest request)
        {
            base._request = request;
            this._messages = this.GenerateMessages();
        }

        public override IEnumerable<Message> GetMessages()
        {
            return this._messages;
        }

        private List<Message> GenerateMessages()
        {
            List<Message> result = new List<Message>();
            var rreq = (ReferenceRequest)base._request;

            var securities = new Dictionary<string, Dictionary<string, object>>();
            foreach (var security in rreq.Securities)
            {
                if (!securities.ContainsKey(security))
                {
                    bool isOption = Regex.IsMatch(security.ToUpper(), @"[A-Z]{1,4}\s+\d{6}[CP]\d{8} EQUITY"); //options look like: AAPL 150117C00600000 EQUITY

                    var fieldData = new Dictionary<string, object>();

                    if (!securities.ContainsKey(security))
                        securities.Add(security, fieldData);
                    else
                    {
                    }

                    List<string> badFields = new List<string>();
                    foreach (var field in rreq.Fields)
                    {
                        object value = Types.RandomDataGenerator.ReferenceDataFromFieldName(field, security, isOption, rreq);
                        if (value != null && !fieldData.ContainsKey(field))
                        {
                            fieldData.Add(field, value);
                        }
                        else
                        {
                        }
                    }
                }
            }

            ReferenceMessage msg = new ReferenceMessage(base._request.correlationId, securities);
            result.Add(msg);

            return result;
        }
    }
}
